capsules/src/sha256.rs

```diff
@@ -0,0 +1,417 @@
+    client: OptionalCell<&'a dyn Client<'a, SHA_256_OUTPUT_LEN_BYTES>>,
+    input_data: OptionalCell<LeasableBuffer<'static, u8>>,
+    data_buffer: MapCell<[u8; SHA_BLOCK_LEN_BYTES]>,
+    buffered_length: Cell<usize>,
+    total_length: Cell<usize>,
+
+    // Used to store the hash or the hash to compare against with verify
+    output_data: Cell<Option<&'static mut [u8; SHA_256_OUTPUT_LEN_BYTES]>>,
+
+    hash_values: Cell<[u32; 8]>,
+    deferred_caller: &'a DynamicDeferredCall,
+    handle: OptionalCell<DeferredCallHandle>,
+}
+
+impl<'a> Sha256Software<'a> {
+    pub fn new(call: &'a DynamicDeferredCall) -> Sha256Software<'a> {
+        Sha256Software {
+            state: Cell::new(State::Idle),
+            client: OptionalCell::empty(),
+            input_data: OptionalCell::empty(),
+            data_buffer: MapCell::new([0; SHA_BLOCK_LEN_BYTES]),
+            buffered_length: Cell::new(0),
+            total_length: Cell::new(0),
+
+            output_data: Cell::new(None),
+            hash_values: Cell::new([0; 8]),
+
+            deferred_caller: call,
+            handle: OptionalCell::empty(),
+        }
+    }
+
+    pub fn initialize_callback_handle(&self, handle: DeferredCallHandle) {
+        self.handle.replace(handle);
+    }
+
+    pub fn busy(&self) -> bool {
+        match self.state.get() {
+            State::Idle => false,
+            _ => true,
+        }
+    }
+
+    pub fn initialize(&self) -> Result<(), ErrorCode> {
+        if !self.busy() {
+            self.buffered_length.set(0);
+            self.data_buffer.map(|b| {
+                for i in 0..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+            });
+            self.hash_values.set([
+                0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab,
+                0x5be0cd19,
+            ]);
+            Ok(())
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+
+    // Complete the hash and produce a final hash result.
+    fn complete_sha256(&self) {
+        let mut buffered_length = self.buffered_length.get();
+        // This shouldn't be necessary, as temp buffer should never be
+        // full. But if it is fill, appending the 1 will be an
+        // out-of-bounds access and panic, so check and clear
+        // the buffered block just in case.
+        if buffered_length == 64 {
+            self.data_buffer.map(|b| {
+                self.compute_block(b);
+                for i in 0..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+            });
+            buffered_length = buffered_length - 64;
+        }
+        if buffered_length < 64 {
+            self.data_buffer.map(|b| {
+                for i in buffered_length..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+            });
+        }
+
+        self.data_buffer.map(|b| {
+            // Append the 1
+            b[buffered_length] = 0x80;
```

Comment:
Why not just use `get()` here and then you don't need the checks above? That also removes a panic :)

Reply:
I don't understand. Can you explain more?

Reply:
We don't do it everywhere in Tock, but for a single allocation like this it's probably a good idea.

You can see an example here: https://doc.rust-lang.org/std/primitive.slice.html#method.get_mut
and Hudson has done a great job explaining it here: https://github.com/tock/tock/blob/master/doc/CodeSize.md#panics

Reply:
Done.